### DISCUSSION TITLE
ILVerify.IResolver with AssemblyName

### DISCUSSION THREAD

**Discussion by KalleOlaviNiemitalo** [dotnet/runtime#64068] [Q&A] [answered]
In the ILVerify.IResolver interface, the Resolve method has only `string simpleName` as a parameter. Can an implementation of this interface get more information about the assembly reference? Preferably the referenced AssemblyName like in [AssemblyLoadContext.Load](https://docs.microsoft.com/dotnet/api/system.runtime.loader.assemblyloadcontext.load?view=net-6.0), and perhaps also a MetadataReader on the referencing module. I would use this to simulate .NET Framework assembly probing when verifying a set of assemblies that support .NET Framework.

<https://github.com/dotnet/runtime/blob/3a25a7f1cc446b60678ed25c9d829420d6321eba/src/coreclr/tools/ILVerification/IResolver.cs#L10-L16>

I found this comment that appears to explain why the AssemblyName is not passed to IResolver, but am not sure what it means.

<https://github.com/dotnet/runtime/blob/3a25a7f1cc446b60678ed25c9d829420d6321eba/src/coreclr/tools/ILVerification/ILVerifyTypeSystemContext.cs#L31-L36>

**Comment by jkotas** [comment, marked as answer]
Yes, I think it should be possible to change this to use `AssemblyName` instead of `string`. 

> perhaps also a MetadataReader on the referencing module

What would you like do with the MetadadaReader of the referencing module?

**Reply by KalleOlaviNiemitalo** [reply]
> Yes, I think it should be possible to change this to use AssemblyName instead of string.

What would that mean for multimodule assemblies? Perhaps it should be Resolve(AssemblyName assemblyName, string moduleName) instead, or a separate method for resolving a module.

What is the breaking-changes policy for IResolver? IIRC, the IL verifier API is still in preview and not yet documented at docs.microsoft.com.

> What would you like do with the MetadadaReader of the referencing module?

I originally thought I could use MetadataReader to look up assembly references, recognise which of them has the same simple name as was passed to IResolver.Resolve, and get the rest of the assembly name from there. However, this won't be necessary if the AssemblyName is already passed in, and would not work if the assembly references multiple versions of another assembly.

Another use would be to read the name of the referencing assembly from the MetadataReader and log that if the referenced assembly or module is not found. [ResolveAssemblyOrNetmodule](https://github.com/dotnet/runtime/blob/5605bdff4b104496941cc9e6acfc19783a52e7b8/src/coreclr/tools/ILVerification/ILVerifyTypeSystemContext.cs#L55-61) can throw an exception in that case, but this exception does not contain any information about the referencing assembly or module.

On the other hand, the PEReader on the referencing module could be more useful than a MetadataReader, as the IResolver could then recognise that it has returned this PEReader before, and use the PEReader as a key in a dictionary that contains related information such as a file path.

**Reply by jkotas** [reply]
> What would that mean for multimodule assemblies? Perhaps it should be Resolve(AssemblyName assemblyName, string moduleName) instead, or a separate method for resolving a module.

I think it would be better to have two different methods: one for assemblies and other for modules of multi-module assemblies.

> What is the breaking-changes policy for IResolver? IIRC, the IL verifier API is still in preview and not yet documented at docs.microsoft.com.

Yes, we do not have a strict compatibility bar for IL verifier APIs.

> ResolveAssemblyOrNetmodule can throw an exception in that case, but this exception does not contain any information about the referencing assembly or module.

We can just fix the exception to have more details. We should not need to pass MetadataReader of referencing module for that.

**Reply by KalleOlaviNiemitalo** [reply]
Like this, then:

```diff
 namespace ILVerify
 {
     public interface IResolver 
     { 
-        PEReader Resolve(string simpleName);
+        PEReader ResolveAssembly(AssemblyName assemblyName);
+        PEReader ResolveModule(PEReader manifestModule, string moduleName);
     }
 }
```

`PEReader manifestModule` would allow relative file lookups for modules of the same assembly, if the resolver remembers the path from which it got each PEReader. (Already in 6.0.0, the resolver needs to remember the PEReader instances that it returns, so that it can return the same ones again, and so that it can eventually close them.)

Does ILVerify resolve the unmanaged modules that are referenced for Platform Invoke? I guess it doesn't, because it does not pass a DllImportSearchPath to the resolver. If it resolved them too, `PEReader manifestModule` could be non-null even when the ModuleRef comes from ImplMap.ImportScope, at least if the module is listed in the File table of the assembly.

If the resolver throws FileNotFoundException, then ILVerify should include the exception instance in the VerificationResult. This would let the resolver report which search paths it used, and whether it found an assembly file with a matching file name but an incompatible assembly version. I don't see a public way for the resolver to set FileNotFoundException.FusionLog, but it could still place the info in Exception.Message or Exception.Data.

**Reply by jkotas** [reply]
The interface shape you have suggested looks good to me.

> Does ILVerify resolve the unmanaged modules that are referenced for Platform Invoke? I guess it doesn't, 

Correct.

**Reply by KalleOlaviNiemitalo** [reply]
If IResolver is changed, then ResolverBase needs changes as well. Minimally:

```diff
 namespace ILVerify
 {
     public abstract class ResolverBase : IResolver 
     {
         public PEReader Resolve(string simpleName);
+        public PEReader ResolveAssembly(AssemblyName assemblyName);
+        public PEReader ResolveModule(PEReader manifestModule, string moduleName);
 
         protected abstract PEReader ResolveCore(string simpleName);
     }
 }
```

Alternatively (but using AssemblyName as the key in ResolverBase._resolverCache would require a custom IEqualityComparer\<T\>, and would be cumbersome if the resolver wants to return the same PEReader for the same file even when the referenced assembly version is different):

```diff
 namespace ILVerify
 {
     public abstract class ResolverBase : IResolver
     {
-        public PEReader Resolve(string simpleName);
+        public PEReader ResolveAssembly(AssemblyName assemblyName);
+        public PEReader ResolveModule(PEReader manifestModule, string moduleName);

-        protected abstract PEReader ResolveCore(string simpleName);
+        protected abstract PEReader ResolveAssemblyCore(AssemblyName assemblyName);
+        protected abstract PEReader ResolveModuleCore(PEReader manifestModule, string moduleName);
     }
 }
```

I have not found ResolverBase useful because it never calls PEReader.Dispose(), and doing that in a derived class requires a separate collection in addition to the ResolverBase._resolverCache that is already there. If I can get the AssemblyName passed in, then the caching by simpleName in ResolverBase will be even less useful. I think it is best to let ResolverBase be used just by the command-line tool and tests, and recommend that more advanced resolvers implement IResolver directly.

Incidentally, should [ResolverBase._resolverCache](https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/coreclr/tools/ILVerification/IResolver.cs#L23) be changed to use StringComparer.OrdinalIgnoreCase? ECMA-335 6th edition section II.22 (Metadata logical format: tables) specifies "case-blind" comparisons of AssemblyRef to Assembly, and ModuleRef to Module. That's also what AssemblyName.ReferenceMatchesDefinition does (although that method has other problems): <https://github.com/dotnet/runtime/blob/ce15dc7d1416ce8ebc4cffe7ea8c51b071f2fd3c/src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyName.cs#L309>

**Reply by jkotas** [reply]
I think it would be fine to delete ResolverBase in the ILVerification.dll library. It has very little value as you have pointed out.

> Incidentally, should ResolverBase._resolverCache be changed to use StringComparer.OrdinalIgnoreCase?

Yes, I agree.

**Reply by KalleOlaviNiemitalo** [reply]
This was pretty much transformed to the issue <https://github.com/dotnet/runtime/issues/65573> so I'll mark the discussion as resolved.
